ios/NebulaNetworkExtension/PacketTunnelProvider.swift

```diff
@@ -132,59 +135,118 @@ class PacketTunnelProvider: NEPacketTunnelProvider {
+            if routeDescription != "" {
+                nebula?.rebind("network change to: \(routeDescription); from: \(cachedRouteDescription ?? "none")")
+            }
+            cachedRouteDescription = routeDescription
+        }
+    }
+    
+    private func collectAddresses(endpoints: [Network.NWEndpoint]) -> String {
+        var str: [String] = []
+        endpoints.forEach{ endpoint in
+            switch endpoint {
+            case let .hostPort(.ipv6(host), port):
+                str.append("[\(host)]:\(port)")
+            case let .hostPort(.ipv4(host), port):
+                str.append("\(host):\(port)")
+            default:
+                return
+            }
+        }
+        
+        return str.sorted().joined(separator: ", ")
     }
     
-    private func wormholeListener(msg: Any?) {
-        guard let call = msg as? IPCRequest else {
+    override func handleAppMessage(_ data: Data, completionHandler: ((Data?) -> Void)? = nil) {
+        guard let call = try? JSONDecoder().decode(IPCRequest.self, from: data) else {
             log("Failed to decode IPCRequest from network extension")
             return
         }
         
         var error: Error?
-        var data: Any?
+        var data: JSON?
+        
+        // start command has special treatment due to needing to call two completers
+        if call.command == "start" {
+            self.start() { error in
+                // Notify the system of our start result
+                if self.startCompleter != nil {
+                    if error == nil {
+                        // Clean boot, no errors
+                        self.startCompleter!(nil)
+                        
+                    } else {
+                        // We encountered an error, we can just pass NSError() here since ios throws it away
+                        // But we will provide it in the event we can intercept the error without doing this workaround sometime in the future
+                        self.startCompleter!(error!.localizedDescription)
+                    }
+                }
+                
+                // Notify the UI if we have a completionHandler
+                if completionHandler != nil {
+                    if error == nil {
+                        // No response data, this is expected on a clean start
+                        completionHandler!(try? JSONEncoder().encode(IPCResponse.init(type: .success, message: nil)))
+                        
+                    } else {
+                        // Error response has
+                        completionHandler!(try? JSONEncoder().encode(IPCResponse.init(type: .error, message: .string(error!.localizedDescription))))
+                    }
+                }
+            }
+            return
+        }
+        
+        if nebula == nil {
+            // Respond with an empty success message in the event a command comes in before we've truly started
```

Comment:
Why success instead of failure? I'm not totally sure what conditions this could occur, but a message seems to imply an action should be taken, and we are not taking that action?

Reply:
The primary purpose right now is to avoid a race where the UI starts issuing commands before core is actually booted, having an error return might make those races more apparent but at the expense of spitting errors at end users and they can't do anything about it.

Reply:
@nbrownus Makes sense - can we log anything, or not easily from here?

Reply:
We can and I added a log statement, nice catch